### PR TITLE
Enable caching for quarkus:build and declare dependencies as inputs

### PR DESCRIPTION
**Pull request by aloubyansky**
@gsmet here are two commits:
1) enhances the `track-config-changes` impl to also dump dependency checksums in `target/quarkus-<profile-name>-dependency-checksums.txt`, typically the profile name would be `prod`;
2) enables dependency checksums dump for all the IT modules.

fyi @jprinet

### CODE REVIEW

File: devtools/maven/src/main/java/io/quarkus/maven/TrackConfigChangesMojo.java [hunk 1]
```diff
@@ -102,34 +112,60 @@ protected void doExecute() throws MojoExecutionException, MojoFailureException {
...
+                        String.class,
+                        Properties.class, QuarkusClassLoader.class, Properties.class, Path.class);
+                dumpConfig.invoke(null, curatedApplication.getApplicationModel(),
+                        launchMode.name(), getBuildSystemProperties(true),
+                        deploymentClassLoader, compareProps, targetFile);
+            }
+
+            if (dumpDependencies) {
+                final List<String> deps = new ArrayList<>();
+                for (var d : curatedApplication.getApplicationModel().getDependencies(DependencyFlags.DEPLOYMENT_CP)) {
```
Comment (gsmet) [inline devtools/maven/src/main/java/io/quarkus/maven/TrackConfigChangesMojo.java:153]: For the extension tests, I wonder if we would have to dump the full test deployment classpath as we might add some deployment extensions for testing.

Comment (aloubyansky) [reply devtools/maven/src/main/java/io/quarkus/maven/TrackConfigChangesMojo.java:153]: That's what it does, unless I misunderstood the point.

File: devtools/maven/src/main/java/io/quarkus/maven/TrackConfigChangesMojo.java [hunk 2]
```diff
@@ -102,34 +112,64 @@ protected void doExecute() throws MojoExecutionException, MojoFailureException {
...
             return;
         }
 
         CuratedApplication curatedApplication = null;
         QuarkusClassLoader deploymentClassLoader = null;
         final ClassLoader originalCl = Thread.currentThread().getContextClassLoader();
-        Properties actualProps;
         final boolean clearPackageTypeSystemProperty = setPackageTypeSystemPropertyIfNativeProfileEnabled();
         try {
             curatedApplication = bootstrapApplication(launchMode);
```
Comment (gsmet) [inline devtools/maven/src/main/java/io/quarkus/maven/TrackConfigChangesMojo.java:126]: FWIW, I benchmarked things a bit and resolving the dependencies is quite slow. Even for a relatively simple IT as the Hibernate Validator one, it takes ~ 0.8 s.

File: devtools/maven/src/main/java/io/quarkus/maven/TrackConfigChangesMojo.java [hunk 3]
```diff
@@ -102,34 +112,64 @@ protected void doExecute() throws MojoExecutionException, MojoFailureException {
...
+                dumpConfig.invoke(null, curatedApplication.getApplicationModel(),
+                        launchMode.name(), getBuildSystemProperties(true),
+                        deploymentClassLoader, compareProps, targetFile);
+            }
+
+            if (dumpDependencies) {
+                final List<String> deps = new ArrayList<>();
+                for (var d : curatedApplication.getApplicationModel().getDependencies(DependencyFlags.DEPLOYMENT_CP)) {
+                    StringBuilder entry = new StringBuilder(d.toGACTVString());
+                    if (d.isSnapshot()) {
```
Comment (gsmet) [inline devtools/maven/src/main/java/io/quarkus/maven/TrackConfigChangesMojo.java:154]: I decided to avoid computing the checksum if the dependency is not a snapshot one. They should be immutable in this case.

File: integration-tests/main/pom.xml [hunk 4]
```diff
@@ -23,6 +23,9 @@
 
         <!-- do not update this dependency, it is only used for testing -->
         <webjar.jquery-ui.version>1.13.0</webjar.jquery-ui.version>
+
+        <quarkus.config-tracking.enabled>true</quarkus.config-tracking.enabled>
+        <quarkus.native.container-build>true</quarkus.native.container-build>
```
Comment (gsmet) [inline integration-tests/main/pom.xml]: Can you clarify why you added that?

File: integration-tests/pom.xml [hunk 5]
```diff
@@ -22,6 +22,9 @@
         <quarkus.build.skip>${skipTests}</quarkus.build.skip>
         <native.surefire.skip>${skipTests}</native.surefire.skip>
         <enforce-test-deps-scope.skip>true</enforce-test-deps-scope.skip>
+
+        <quarkus.config-tracking.enabled>true</quarkus.config-tracking.enabled>
+        <quarkus.native.container-build>true</quarkus.native.container-build>
```
Comment (gsmet) [inline integration-tests/pom.xml:27]: Can you clarify why you added `<quarkus.native.container-build>true</quarkus.native.container-build>`?

**Review by gsmet — COMMENTED** [review]
Things seem to work with the changes I made.

Could you clarify again why it only works the third time? I somehow forgot about it.

**Comment by gsmet** [comment]
BTW, it wasn't terribly clear but I added a bunch of commits.

File: integration-tests/pom.xml (hunk 5 above)
Comment (jprinet) [reply integration-tests/pom.xml:27]: This is required by the [Quarkus caching extension](https://github.com/gradle/develocity-build-config-samples/tree/main/quarkus-build-caching-extension#limitations) to ensure native builds are as reproducible as possible and make caching consistent.

**Comment by jprinet** [comment]
> Things seem to work with the changes I made.
> 
> Could you clarify again why it only works the third time? I somehow forgot about it.

This is not true anymore, with the config dump [cached/restored](https://github.com/quarkusio/quarkus/pull/37622/files#diff-d8e36253822cffbdf21e333bcba467345b46e6095927dda98399fe7f21f4679aR1032) on the CI runner, the cache hit would occur on first run if inputs are identical.

You can check the initial explanation [here](https://github.com/quarkusio/quarkus/pull/37258#issuecomment-1827479640)

File: devtools/maven/src/main/java/io/quarkus/maven/TrackConfigChangesMojo.java (hunk 3 above)
Comment (aloubyansky) [reply devtools/maven/src/main/java/io/quarkus/maven/TrackConfigChangesMojo.java:154]: Final versions could still be built locally. For example before releasing but not necessarily then.

File: devtools/maven/src/main/java/io/quarkus/maven/TrackConfigChangesMojo.java (hunk 2 above)
Comment (aloubyansky) [reply devtools/maven/src/main/java/io/quarkus/maven/TrackConfigChangesMojo.java:126]: It is, this is something i've been looking into for the last few weeks. But there is not much we can do especially for relatively simple projects, since most of the time is spent in the Maven artifact resolver itself.

However, the app model resolution will typically be necessary either way. It's about at which point it's performed. Here, we are moving this step to an earlier phase and it's cached for later phases if they get executed.

File: devtools/maven/src/main/java/io/quarkus/maven/TrackConfigChangesMojo.java [hunk 6]
```diff
@@ -75,7 +75,7 @@ public class TrackConfigChangesMojo extends QuarkusBootstrapMojo {
     /**
      * Dependency dump file
      */
-    @Parameter(property = "quarkus.track-config-changes.dependenciesFile")
```
Comment (aloubyansky) [inline devtools/maven/src/main/java/io/quarkus/maven/TrackConfigChangesMojo.java]: I suppose to be completely consistent, it would have to be `quarkus.trackConfigChanges.dependenciesFile`

**Review by gsmet — APPROVED** [review]

File: integration-tests/main/pom.xml (hunk 4 above)
Comment (zakkak) [inline integration-tests/main/pom.xml:28]: @rsvoboda why is this change necessary?

Comment (jprinet) [reply integration-tests/main/pom.xml:28]: This is required to make the Quarkus build goal cacheable:
https://github.com/gradle/develocity-build-config-samples/tree/main/quarkus-build-caching-extension

Comment (zakkak) [reply integration-tests/main/pom.xml:28]: Good, thanks for the pointer @jprinet
